### PR TITLE
treat SSL connection that doesn't terminate SSL but gracefully termin…

### PR DESCRIPTION
**Pull request by rdp**
…ates the underlying connection as an EOF #7383

### CODE REVIEW

File: src/openssl/ssl/socket.cr [hunk 1]
```diff
@@ -115,7 +115,13 @@ abstract class OpenSSL::SSL::Socket < IO
 
     LibSSL.ssl_read(@ssl, slice.to_unsafe, count).tap do |bytes|
       if bytes <= 0 && !LibSSL.ssl_get_error(@ssl, bytes).zero_return?
-        raise OpenSSL::SSL::Error.new(@ssl, bytes, "SSL_read")
+        ex = OpenSSL::SSL::Error.new(@ssl, bytes, "SSL_read")
+        if ex.underlying_eof?
+          # underlying BIO termianted gracefully, without terminating SSL aspect gracefully first
```
Comment (Sija) [inline src/openssl/ssl/socket.cr]:
```suggestion
          # underlying BIO terminated gracefully, without terminating SSL aspect gracefully first
```

**Comment by rdp** [comment]
OK this causes a hang in the unit tests on Linux, investigating...

**Comment by rdp** [comment]
This is basically kind of blocked until #8582 is resolved

**Comment by rdp** [comment]
OK ready for review now, the unit tests pass in Linux now :)

**Comment by bcardiff** [comment]
@rdp I'm afraid I need to ask for a rebase on master. The PR currently shows lots of other changes altogether.

**Comment by rdp** [comment]
OK should be ready for review

**Pull request merged**